core/org.eclipse.cdt.core/parser/org/eclipse/cdt/internal/core/dom/parser/cpp/semantics/CPPSemantics.java

```diff
@@ -409,6 +409,24 @@ private static IBinding postResolution(IBinding binding, LookupData data) {
 			}
 		}
 
+		// If any name in the initializer refers to introduced names then declaration is ill-formed.
+		if (binding instanceof ICPPInternalBinding && ((ICPPInternalBinding) binding).getDefinition() != null) {
+			IASTNode definitionParent = ((ICPPInternalBinding) binding).getDefinition().getParent();
+			if (definitionParent instanceof ICPPASTStructuredBindingDeclaration) {
+				ICPPASTStructuredBindingDeclaration declaration = (ICPPASTStructuredBindingDeclaration) definitionParent;
```

Comment:
Just my personal taste, for this kind of if statements pattern matching comes in handy
```suggestion
		if (binding instanceof ICPPInternalBinding internalBinding && internalBinding.getDefinition() != null
				&& internalBinding.getDefinition()
						.getParent() instanceof ICPPASTStructuredBindingDeclaration declaration) {
```